kafka-rest/src/main/java/io/confluent/kafkarest/resources/v2/PartitionsResource.java

```diff
@@ -200,13 +216,124 @@ public void onCompletion(
+   * its associated timestamp, or empty if no such offset exists.</p>
+   *
+   * @throws io.confluent.rest.exceptions.RestNotFoundException if either {@code topic} or {@code
+   *                                                            partition} don't exist.
+   */
+  @GET
+  @Path("/{partition}/offsets")
+  public GetTopicPartitionOffsetResponse getOffsets(
+      @PathParam("topic") String topic,
+      @PathParam("partition") int partition,
+      @QueryParam("timestamp") @Nullable Instant timestamp
+  ) throws Exception {
+    checkTopicExists(topic);
+    checkPartitionExists(topic, partition);
+
+    if (timestamp == null) {
+      return GetTopicPartitionOffsetResponse.createSummary(
+          getBeginningOffset(topic, partition), getEndOffset(topic, partition));
+    }
+
+    return getOffsetForTime(topic, partition, timestamp)
+        .map(
+            offsetTimestamp ->
+                GetTopicPartitionOffsetResponse.createForTime(
+                    offsetTimestamp.getOffset(), offsetTimestamp.getTimestamp()))
+        .orElse(GetTopicPartitionOffsetResponse.empty());
+  }
+
+  /**
+   * Returns the earliest offset in the {@code topic} {@code partition}.
+   */
+  private long getBeginningOffset(String topic, int partition) {
+    try {
+      return retrying(
+          () -> ctx.getKafkaConsumerManager()
+              .getBeginningOffset(consumerGroup, getConsumerInstance(), topic, partition),
+          ConsumerInstanceNotFoundException.class,
+          CONSUMER_REQUEST_MAX_ATTEMPTS);
+    } catch (ConsumerInstanceNotFoundException e) {
+      throw new IllegalStateException(e);
+    }
+  }
+
+  /**
+   * Returns the latest offset in the {@code topic} {@code partition}.
+   */
+  private long getEndOffset(String topic, int partition) {
+    try {
+      return retrying(
+          () -> ctx.getKafkaConsumerManager()
+              .getEndOffset(consumerGroup, getConsumerInstance(), topic, partition),
+          ConsumerInstanceNotFoundException.class,
+          CONSUMER_REQUEST_MAX_ATTEMPTS);
+    } catch (ConsumerInstanceNotFoundException e) {
+      throw new IllegalStateException(e);
+    }
+  }
+
+  /**
+   * Returns the earliest offset in the {@code topic} {@code partition} whose timestamp is after the
+   * given {@code timestamp} and its associated timestamp, or empty if no such offset exists.</p>
+   */
+  private Optional<OffsetTimestamp> getOffsetForTime(
+      String topic, int partition, Instant timestamp) {
+    try {
+      return retrying(
+          () -> ctx.getKafkaConsumerManager()
+              .getOffsetForTime(consumerGroup, getConsumerInstance(), topic, partition, timestamp),
+          ConsumerInstanceNotFoundException.class,
+          CONSUMER_REQUEST_MAX_ATTEMPTS);
+    } catch (ConsumerInstanceNotFoundException e) {
+      throw new IllegalStateException(e);
+    }
+  }
+
+  private String createConsumerGroup() {
```

Comment:
i'm thinking incorporate this in consumer manager / invent the concept of a 'utility consumer' or something if there's no such thing already.